market-generator/src/main/java/com/market/banica/generator/service/MarketStateImpl.java

```diff
@@ -0,0 +1,124 @@
+package com.market.banica.generator.service;
+
+import com.market.TickResponse;
+import com.market.banica.generator.model.MarketTick;
+import com.market.banica.generator.util.SnapshotPersistence;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
+
+@Component
+public class MarketStateImpl implements com.market.banica.generator.service.MarketState {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MarketStateImpl.class);
+
+    private static final ReadWriteLock marketStateLock = new ReentrantReadWriteLock();
+
+    private static final SnapshotPersistence snapshotPersistence = new SnapshotPersistence();
+
+    private final Map<String, Set<MarketTick>> marketState;
+
+    private final ExecutorService executorService;
+
+    private final MarketSubscriptionManager subscriptionManager;
+
+    @Autowired
+    public MarketStateImpl(MarketSubscriptionManager subscriptionManager) throws IOException {
+        this.marketState = snapshotPersistence.loadPersistedSnapshot();
+        this.executorService = Executors.newSingleThreadExecutor();
+        this.subscriptionManager = subscriptionManager;
+        PersistScheduler persistScheduler = new PersistSchedulerImpl(marketStateLock, snapshotPersistence, marketState);
+        persistScheduler.scheduleSnapshot();
+    }
+
+
+    @Override
+    public void addTickToMarketState(MarketTick marketTick) {
+        executorService.execute(() -> {
```

Comment:
why is this executed in executor ?

Reply:
The idea is that whenever the Timer thread invokes addTickToMarketState method it passes the work to a different thread, since the Timer has only 1 working thread and it is used only to generate new ticks, if we do not pass the work to a different thread it will slow the whole app significantly.

Reply:
Ok but your executor is initialized with an Executors.newSingleThreadExecutor. This means it will always use just 1 thread, which from my point of view doesn't help for accelerating the app.

Reply:
I think that Executors.newSingleThreadExecutor internally adds immediately all the passed tasks in a queue and executes them one by one in its single thread, this way it shouldn't slow down the Timer thread. I am ok with this approach